ksmt-z3/src/main/kotlin/io/ksmt/solver/z3/KZ3ForkingSolverManager.kt

```diff
@@ -0,0 +1,172 @@
+package io.ksmt.solver.z3
+
+import com.microsoft.z3.Context
+import com.microsoft.z3.Z3Exception
+import com.microsoft.z3.decRefUnsafeAll
+import io.ksmt.KAst
+import io.ksmt.KContext
+import io.ksmt.decl.KDecl
+import io.ksmt.expr.KExpr
+import io.ksmt.expr.KUninterpretedSortValue
+import io.ksmt.solver.KForkingSolver
+import io.ksmt.solver.KForkingSolverManager
+import io.ksmt.solver.KSolverException
+import io.ksmt.solver.util.KExprLongInternalizerBase
+import io.ksmt.sort.KSort
+import io.ksmt.sort.KUninterpretedSort
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap
+import it.unimi.dsi.fastutil.longs.LongOpenHashSet
+import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap
+import java.util.IdentityHashMap
+import java.util.concurrent.ConcurrentHashMap
+
+class KZ3ForkingSolverManager(private val ctx: KContext) : KForkingSolverManager<KZ3SolverConfiguration> {
+    private val solvers = ConcurrentHashMap.newKeySet<KZ3ForkingSolver>()
+
+    /**
+     * for each parent-to-child hierarchy created only one Context.
+     * Each Context user is registered to control solver is alive
+     */
+    private val forkingSolverToContext = IdentityHashMap<KZ3ForkingSolver, Context>()
+    private val contextReferences = IdentityHashMap<Context, Int>()
```

Comment:
If we always use the same context what is the reason to have identity map from context to something? The key will always be the same.

Reply:
We can create root solvers via manager on `mkForkingSolver()` without params. Each of the root produces a new `Context`, which is reusable for vertical forking hierarchy, not for horizontal one. That's why one manager is able to have many `Context`-s